core/src/splitting/MultiRegressionSplittingRule.cpp

```diff
@@ -0,0 +1,205 @@
+/*-------------------------------------------------------------------------------
+  This file is part of generalized random forest (grf).
+
+  grf is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  grf is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with grf. If not, see <http://www.gnu.org/licenses/>.
+ #-------------------------------------------------------------------------------*/
+
+#include <algorithm>
+
+#include "MultiRegressionSplittingRule.h"
+
+namespace grf {
+
+MultiRegressionSplittingRule::MultiRegressionSplittingRule(size_t max_num_unique_values,
+                                                           double alpha,
+                                                           double imbalance_penalty,
+                                                           size_t num_outcomes):
+    alpha(alpha),
+    imbalance_penalty(imbalance_penalty),
+    num_outcomes(num_outcomes) {
+  this->counter = new size_t[max_num_unique_values];
+  this->sums = Eigen::ArrayXXd(max_num_unique_values, num_outcomes);
+  this->weight_sums = new double[max_num_unique_values];
+}
+
+MultiRegressionSplittingRule::~MultiRegressionSplittingRule() {
+  if (counter != nullptr) {
+    delete[] counter;
+  }
+  if (weight_sums != nullptr) {
+    delete[] weight_sums;
+  }
+}
+
+bool MultiRegressionSplittingRule::find_best_split(const Data& data,
+                                                   size_t node,
+                                                   const std::vector<size_t>& possible_split_vars,
+                                                   const Eigen::ArrayXXd& responses_by_sample,
+                                                   const std::vector<std::vector<size_t>>& samples,
+                                                   std::vector<size_t>& split_vars,
+                                                   std::vector<double>& split_values,
+                                                   std::vector<bool>& send_missing_left) {
+
+  size_t size_node = samples[node].size();
+  size_t min_child_size = std::max<size_t>(std::ceil(size_node * alpha), 1uL);
```

Comment:
Just a note, all the internals here are practically the same as the single-variable splitting rule, but for multi-treatment causal forest (follow-up PR) I would like to replace this line with:

`size_t min_child_size = std::max<size_t>(std::ceil(size_node * alpha), min_node_size);`

in order to side-step adding all kinds of split constraints in a new splitting rule (minimum number of treatments on each side, etc etc), and instead enforce a reasonably large node size.